solvers/bcd_celer.py

```diff
@@ -0,0 +1,363 @@
+                    verbose):
+    K = U.shape[0] + 1
+
+    if epoch // gap_freq < K:
+        last_K_R[(epoch // gap_freq), :] = R
+    else:
+        for k in range(K - 1):
+            last_K_R[k, :] = last_K_R[k + 1, :]
+        last_K_R[K - 1, :] = R
+        for k in range(K - 1):
+            U[k] = last_K_R[k + 1].ravel() - last_K_R[k].ravel()
+
+        # double for loop but small: K**2/2
+        for k in range(K - 1):
+            for j in range(k, K - 1):
+                UtU[k, j] = np.dot(U[k], U[j])
+                UtU[j, k] = UtU[k, j]
+
+        try:
+            anderson = np.linalg.solve(UtU, np.ones(UtU.shape[0]))
+        except Exception:
+            # np.linalg.LinAlgError
+            # Numba only accepts Error/Exception inheriting from the generic
+            # Exception class
+            if verbose:
+                print("Singular matrix when computing accelerated point.")
+        else:
+            anderson /= np.sum(anderson)
+
+            out[:] = 0
+            for k in range(K - 1):
+                out += anderson[k] * last_K_R[k, :]
+            out /= alpha
+            # out now holds the extrapolated dual point
+
+
+def bcd_epoch(C, norms_X_block, X, R, alpha, W, inv_lc):
+    n_tasks = R.shape[1]
+    W_j_new = np.zeros((1, n_tasks))
+    dgemm = _get_dgemm()
+    alpha_lc = alpha * inv_lc
+
+    for j in C:
+        if norms_X_block[j] == 0.:
+            continue
+        idx = slice(j, j+1)
+        W_j = W[idx, :]
+        X_j = X[:, idx]
+
+        # W_j_old = W[j].copy()
+
+        # W_j_new = X_j.T @ R * inv_lc[j]
+        dgemm(alpha=inv_lc[j], beta=0.0, a=R.T, b=X_j, c=W_j_new.T,
+              overwrite_c=True)
+
+        if W_j[0, 0] != 0:
+            # R += np.dot(X_j, W_j)
+            dgemm(alpha=1.0, beta=1.0, a=W_j.T, b=X_j.T, c=R.T,
+                  overwrite_c=True)
+            W_j_new += W_j
+
+        # block_norm = np.sqrt(sum_squared(W_j_new))
+        block_norm = norm(W_j_new)
+
+        if block_norm <= alpha_lc[j]:
+            W_j.fill(0.0)
+        else:
+            shrink = 1.0 - alpha_lc[j] / block_norm
+            W_j_new *= shrink
+            # R -= np.dot(X_j, W_j_new)
+            dgemm(alpha=-1.0, beta=1.0, a=W_j_new.T, b=X_j.T, c=R.T,
+                  overwrite_c=True)
+            W_j[:] = W_j_new
+
+        # R += np.outer(X[:, j], W_j_old - W[j])
+        # R[:] = Y - X @ W
+        # np.testing.assert_allclose(Y - X @ W, R)
+
+
+def celer_dual_mtl(X, Y, alpha, n_iter, max_epochs=10_000, gap_freq=10,
+                   tol=1e-12, p0=10, verbose=0, prune=True, accel=True):
+    n_samples, n_features = X.shape
+    n_tasks = Y.shape[1]
+    n_obs = n_samples * n_tasks
+
+    W = np.zeros((n_features, n_tasks))
+    R = Y.copy()
+    verbose_in = max(verbose - 1, 0)
+    n_screened = 0
+    norm_Y2 = norm(Y, 'fro') ** 2
+
+    tol *= norm_Y2
+    if p0 > n_features:
+        p0 = n_features
+
+    prios = np.empty(n_features, dtype=X.dtype)
```

Comment:
it's not worth using np.empty instead of np.zeros, this can lead to hard to debug errors from my experience